### ISSUE TITLE
[Epic] Sequential dataflow hydration

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#25271]
Hydration spikes are often caused and exacerbated by the fact that newly created compute replicas start hydrating all installed dataflows immediately and in parallel. This requires the replica to handle the increased memory usage that comes from processing the (likely large) source snapshots from all dataflows simultaneously. Replicas might have enough spare memory to hydrated each individual dataflow, but not enough to hydrate all at the same time. 

## Product outcome

* Hydration memory spikes are significantly reduced on replicas that have multiple dataflows installed.

## Discovery

We have validated that sequential hydration can improve overall hydration speed when lgalloc is enabled (https://github.com/MaterializeInc/materialize/issues/25271#issuecomment-1964513933). This is likely because it reduces peak memory usage and therefore the amount of data that is written to disk. Ooming is less of an issue with lgalloc but can still occur when a replica runs out of disk space.

There are still open questions to answer:

* Do we want fully sequential hydration, or just reduced parallelism?
* (How) Should hydration parallelism be configurable by users?
* Should sequential hydration be applied only to "initial" dataflows or also dataflows installed during steady state?
* Should sequential hydration be synchronized between workers?

## Work items

```[tasklist]
- [ ] #1
```

## Decision log

* **15 February 2024.** Issue created.

**Comment by teskje** [comment]
Test results with a TPC-H workload (SF 10, indexes on all 22 queries, no pk/fk indexes) and various lgalloc configurations:

* `concurrency` is the amount of dataflows that are permitted to hydrate at the same time.
* `ru_maxrss` is the maxrss value from [getrusage](https://man7.org/linux/man-pages/man2/getrusage.2.html). Contains both spillable and non-spillable memory.
* `working set` is the highest peak in the `container_memory_working_set_bytes` metric. Contains only non-spillable memory (I think). Due to sampling, the actual max value was probably higher.

### No lgalloc

concurrency |   ru_maxrss | working set | hydration time
------------|-------------|-------------|---------------
1 |    53.8 GiB |    45.9 GiB |        18 min
2 |    60.1 GiB |    58.5 GiB |        17 min
3 |    68.6 GiB |    58.4 GiB |        16 min
4 |    63.1 GiB |    60.4 GiB |        15 min
10 |   177.0 GiB |   166.0 GiB |        15 min
inf |       (OOM) |       (OOM) |         (OOM)

### Full lgalloc (with persist_source)

concurrency |   ru_maxrss | working set | hydration time
------------|-------------|-------------|---------------
1 |    73.0 GiB |    19.5 GiB |        27 min
2 |    99.9 GiB |    22.0 GiB |        21 min
3 |   101.0 GiB |    24.0 GiB |        22 min
4 |    97.2 GiB |    27.6 GiB |        23 min
10 |   188.0 GiB |    35.8 GiB |        34 min
inf |   228.0 GiB |    66.5 GiB |        36 min

### Partial lgalloc (no persist_source)

concurrency |   ru_maxrss | working set | hydration time
------------|-------------|-------------|---------------
1 |    95.1 GiB |    19.4 GiB |        20 min
2 |   116.0 GiB |   110.0 GiB |        18 min
3 |   137.0 GiB |   122.0 GiB |        17 min
4 |   168.0 GiB |   154.0 GiB |        26 min
10 |       (OOM) |       (OOM) |         (OOM)
inf |       (OOM) |       (OOM) |         (OOM)

We see that the amount of concurrency influences both memory usage and hydration time. Generally:
* The higher the concurrency, the higher the memory usage.
* The higher the concurrency, the shorter the hydration time.

But: With lgalloc, hydration time does get worse pretty quickly, at a concurrency of 4 in this workload. My theory is that the higher memory usage makes us spill to disk more, causing more slow disk operations.

The test indicates some benefit of sequential hydration for both memory usage (which we care less about in a world with lgalloc) and hydration time (which we care more about).

**Comment by teskje** [comment]
# Implementation considerations

Implementing this feature will likely be harder than we initially expected.

First, we have the requirement to not degrade replica responsiveness during hydration. Doing so would throw us back into the bad state where we can't debug hydration bottlenecks and where users see a bunch of errors in the console while their dataflows hydrate. That means we cannot simply hydrate all dataflows in creation order, we need to let some dataflows (for slow-path `SELECT`s and probably also `SUBSCRIBE`s) jump the queue.

Timely requires that all workers in a multi-worker cluster render dataflows in the same order. Rendering some dataflows sequentially and others immediately makes this hard to ensure. Consider this scenario with two workers and `concurrency = 1`:

worker 1 | worker 2
---------|---------
receive: `CreateDataflow(u1)` | receive: `CreateDataflow(u1)`
render:  `u1` | render: `u1`
receive: `CreateDataflow(u2)` | receive: `CreateDataflow(u2)`
enqueue: `u2` | enqueue: `u2`
hydrated: `u1` |
render: `u2` |
receive: `CreateDataflow(t1)` | receive: `CreateDataflow(t1)`
render: `t1` | render: `t1`
|| hydrated: `u1`
|| render: `u2`

This assumes that `t1` is a transient dataflow that should be rendered immediately while `u1` and `u2` are long-lived dataflows that need to adhere to `concurrency`. The order of dataflow commands is guaranteed to be consistent across workers, but there are no guarantees about how hydration events interleave with the command stream and therefore also none about when workers render previously enqueued dataflows. In the example, we end up rendering `u1`-`u2`-`t1` on worker 1 and `u1`-`t1`-`u2` on worker 2, which will probably crash Timely.

A plausible way to work around this is to not delay dataflow rendering, but rather delay compute commands at a point before they are partitioned by worker. We can imagine a new `SequentialHydration` client that sits before the controller-side `PartitionedState` and holds back `CreateDataflow` commands to achieve the desired `concurrency`. Such a client would release more `CreateDataflow` commands in response to `FrontierUpper` responses that announce hydration of previously created dataflows.

This unfortunately doesn't work with reconciliation: After a reconnection, a replica will receive dataflow commands and stash them away without executing them until it sees an `InitializationComplete` command. Only then does it reconcile its state with the received commands and starts rendering new dataflows. The `SequentialHydration` client has no correct way to work around this. If it holds back the `InitializationComplete` command waiting for `FrontierUpper` responses, it enters into a deadlock with the replica that waits for the `InitializationComplete` command. If it does not hold back the `InitializationCommand`, it breaks reconciliation for all dataflows that it holds back because the replica will assume that its hydrated instances of these dataflows are no longer needed.

The way reconciliation is currently implemented means that the decisions about sequential dataflow hydration have to happen after it in the pipeline. Reconciliation is performed per-worker, so sequential dataflow decisions must be made by worker too. This conflicts with the need to ensure a consistent order of dataflow creations among all workers.

Some ideas, neither of them particularly straightforward to implements:

* Render all dataflows as `CreateDataflow` commands arrive. Teach source operators (`persist_source`, arrangement imports) to hold producing any data until a token is dropped. Let the `ComputeState` drop the token to control which dataflows may begin hydration.
* Render all dataflows as `CreateDataflow` commands arrive. Add a Timely API that allows "deactivate" dataflows so they don't get scheduled. Let the `ComputeState` activate dataflows to control which ones may begin hydration.
* Introduce some shared state between workers that they can use to agree which dataflow should be rendered next.

Note that any approach that doesn't render dataflows as their `CreateDataflow` commands arrive has to also be concerned with holding back `AllowCompaction` commands. Consider this sequence of commands:
* `CreateDataflow(u1, "SELECT 1")`
* `CreateDataflow(u2, "SELECT * FROM u1")`
* `AllowCompaction(u1, [])`

If rendering of `u2` is held back and the `AllowCompaction` command is executed first, `u1` may get dropped without `u2` being able to obtain a read hold on it. Later, rendering of `u2` would fail because `u1` doesn't exist anymore. To avoid this, the `AllowCompaction` command needs to be held back until `u2` has been rendered, or there must be some other mechanism to ensure that `u2` read capability on `u1` is remembered.

**Comment by teskje** [comment]
The current thinking is that we should always immediately render dataflows, but then somehow suspend those that shouldn't hydrate yet until we allow them to hydrate. This works around all the implementation issues above.

For now we are focussing on ways that don't require changing Timely. That rules out adding an API to Timely to suspend dataflows, and also rules out providing a suspension token to arrangement import operators. But we can modify `persist_source` to take a suspension token. Doing so will not be as complete of a solution as suspending all dataflow sources, but perhaps one that still yields benefits.

To validate this suspicion, I ran a test with an implementation of sequential hydration that suspends `persist_source`s only. This test uses [the previous TPC-H workload](https://github.com/MaterializeInc/materialize/issues/25271#issuecomment-1964513933) but I only collected results in the "Full lgalloc" configuration, as this is also the one we eventually want to run in production. Out of curiosity I ran the test on both a `large` and a `medium` replica. These are the results:

* `large` replica:

concurrency |   ru_maxrss | working set | hydration time
------------|-------------|-------------|---------------
1 |    51.7 GiB |    38.9 GiB |        17 min
2 |    54.3 GiB |    61.4 GiB |        16 min
3 |    61.2 GiB |    61.3 GiB |        18 min
4 |    68.8 GiB |    61.4 GiB |        18 min
10 |   112.0 GiB |   113.0 GiB |        25 min
inf |   181.0 GiB |   182.0 GiB |        27 min

* `medium` replica:

concurrency |   ru_maxrss | working set | hydration time
------------|-------------|-------------|---------------
1 |    48.4 GiB |    42.2 GiB |        24 min
2 |    49.8 GiB |    54.9 GiB |        25 min
3 |    46.7 GiB |    57.5 GiB |        26 min
4 |    49.6 GiB |    64.0 GiB |        27 min
10 |    92.3 GiB |   107.0 GiB |        31 min
inf |    96.3 GiB |    44.8 GiB |        34 min

The main observation is that sequential hydration is still useful for reducing the hydration memory spike and thus for reducing hydration time. That is actually not a surprising outcome, as the chosen TPC-H workload only runs dataflows that read from persist directly, not from indexes. We can expect the results to be worse for workloads that have dataflows reading from indexes.
